Build pipelining support

eed3si9n opened the pull request:
This is a work-in-progress, open source port of build pipelining support implemented by Lightbend Scala team (https://github.com/scala/scala/pull/7712) and Morgan Stanley (https://github.com/mspnf/zinc/blob/1.3.x-contrib/README-MORGANSTANLEY.MD).

- [x] Macro support
- [x] ~Java support~
- [x] sbt support

Ref scala/scala#7712
Ref scala/scala-dev#643

This adds a new flag `incOption.withPipelining(true)` that enables incremental build pipelining.

**Note**: If incOptions pipelining is enabled, the build tool must call `compileAllJava(in, logger)` after _both_
1. All upstream subproject Scala and Java compilation has completed, and 
2. Scala *whole* `*.class` compilation has completed && hasModified.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/744/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eed3si9n committed:
Expand CompileProgress and Setup

Early analysis store is passed into the Setup.
CompileProgress is used to notify the early output timing.

eed3si9n committed:
Handle macros

If a macro is found in any classes, use the regular output.

eed3si9n committed:
Use -Ypickle-java to generate pickles from Java

dwijnand commented:
@eed3si9n any chance you can rebase this one?  It's merge-conflicting with other changes (like the TestProjectSetup reflector).

eed3si9n commented:
@dwijnand Sure. I guess I can split ready-to-go pieces apart while I'm at it.

dwijnand commented:
Yeah, that sounds like a great idea.

eed3si9n committed:
Implement pipelining using custom pickle jar

The deletion part required some tweaking to make sure this works for dir-based output.

eed3si9n committed:
Add pipelining to IncOptions

eed3si9n committed:
Implement early output and early analysis

For modular pipelining we need both early output (JAR file containing Scala sig files) and early Analysis (Zinc internal information).

This adds `IncOptions#earlyOutput` and `Lookup#storeEarlyAnalysis` so the early artifacts can be generated during compile phases.

retronym commented:
It would be good if we could use `PathFactory` and  `-Ypickle-write` to get the pickles out of the compiler, rather than having the new `pickleData` extension point in the compiler callback.

Here's what I had in mind:

https://github.com/scala/scala/blob/8538d98989b33d0090fdd142435320abae10a2fd/test/junit/scala/tools/nsc/PickleWriteTest.scala#L52-L97

This can be combined with `-Ypickle-write-api-only` which limits the written `.sig` files to non-private API. Such pickles aren't (currently) stored in a map like `Run#symData`, as I wanted to avoid a) another ad-hoc extension point and b) keeping public pickles in memory until the end of the `Run`.

We can stick with this `Run.symData` to get to a viable version of this PR though.

eed3si9n commented:
Here are some challenges:

1. keeping `*.class` and `*.sig` in sync, i.e. propagating deletion of `*.class` to `*.sig`,
2. backing out `A.class` and `A.sig` while compiling `A.scala`,
3. supporting both compile-to-JAR and -to-directory,
4. making sure it works for both Java and Scala, and
5. making sure it works for subproject dependencies.

For example, given subprojects core (`A.scala`, ...) and app (`App.scala`) that depends on core, transactional `ClassFileManager` today backs out `A.class` when compiling `A.scala`.

1. Compile both app and core subprojects.
2. Introduce a change to `A.scala` so it no longer compiles.
3. Try compiling app and core again. `A.sig` would still exist in the classpath, so `App.scala` compilation succeeds (To be clear that's a problem. The compilation of `App.scala` *should* fail).

smarter commented:
> keeping *.class and *.sig in sync, i.e. propagating deletion of *.class to *.sig,

We have the exact same problem with .tasty, which is why we only emit .tasty in the backend even though they could be emitted at pickler (and even then we also need all of this wonderful stuff: https://github.com/lampepfl/dotty/blob/11760ea2b23ad7eb5b5749f489ee5395980c0daf/sbt-dotty/src/dotty/tools/sbtplugin/DottyPlugin.scala#L127-L152)

eed3si9n commented:
Also to document "making sure it works for both Java and Scala" part of the challenge... Here's `MixedAnalyzingCompiler#compile` as it stands:

https://github.com/sbt/zinc/blob/0a961166a17e25889d08677b7adf4117714cc5fc/zinc/src/main/scala/sbt/internal/inc/MixedAnalyzingCompiler.scala#L157-L163

By default it calls `compileScala()` and immediately afterwards it calls `compileJava()`.

### delay Javac compilation?

Using core and app subproject as example, let's say both subprojects are Scala and Java sources. The question is when can we start "app" subproject compile. Ideally we want to start as soon as early artifacts are available from core. `compileScala()` should go through fine. `compileJava()` would fail because Javac wants real `*.class` file in the classpath.

So should we wait till whole compilation is done for core subproject? Doing so would undermine the pipelining whenever `*.java` files are included. If we want to avoid that, we need a way to queue Javac compilation, and wait until whole compilations of subproject dependencies are done.

### *.class-based JavaAnalyze?

AnalyzingJavaCompiler (https://github.com/sbt/zinc/blob/v1.4.0-M5/zinc/src/main/scala/sbt/internal/inc/javac/AnalyzingJavaCompiler.scala) runs analysis on top of the `*.class` file created by `javac`.

If we try to pipeline mixed Java/Scala subprojects, the Analysis information created by `JavaAnalyze` will not be available to downstream subprojects. Instead it will use the Analysis information created via `-Ypickle-java` + xsbt-dependency phase etc. I'm not sure if we lose some accuracy because of this.

eed3si9n commented:
Here's a diagram to illustrate this.

![pipelining](https://user-images.githubusercontent.com/184683/81020232-58f43900-8e36-11ea-8239-3bacb0452a8b.png)

retronym commented:
> So should we wait till whole compilation is done for core subproject?

The intention is that we use `.sig` files generated by `-Ypickle-java` as inputs to downstream `scalac`, and `.class` files as inputs to downstream `javac`.`app` is too coarse a node in this dependency graph, we need `app:scalac` and `app:java` as separate nodes.

Downstream projects that use `.sig` files will be presented with an partial ("early"?) `Analysis` that lacks some information. As you pointed out, `AnalyzingJavaCompiler` currently extracts the API, used libraries, used internal classes from `.class` files. 

AFAICT some of this information (e.g. external library dependencies) is irrelevant for the decision making of the downstream incremental compiler and won't be needed until subsequent compilation of this project. The API extraction _is_ important, but should be safe to replace with `extact-api` processing the API extracted by `scalac -Ypickle-java`. The API hash won't be identical due to implementation differences, but that's not a problem if you're _always_ using one strategy or the other.

I think it would be useful to validate our assumptions about what information is irrelevant to downstream incremental compilation, they would only be consulted as part of `previousRelations`. Could we validated the assumptions [by modifying `LookupImpl.analyses` ](https://github.com/sbt/zinc/compare/a266b8c4c7a068c722a00e6f95486e401925005f...retronym:topic/classpath-analysis?expand=1)to filter out the information we believe is not needed from classpath analyses?

eed3si9n commented:
> Could we validated the assumptions by modifying LookupImpl.analyses to filter out the information we believe is not needed from classpath analyses?

Good idea. I think all we need are `analysis.relations.productClassName` and `analysis.apis.internalAPI` accessed by https://github.com/sbt/zinc/blob/0a961166a17e25889d08677b7adf4117714cc5fc/internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala#L776-L783

`analysis.relations.productClassName` is determined by `nonLocalClasses` and `localClasses` collected in `AnalysisCallback`. In API Extraction phase `registerGeneratedClasses` registers non-local `*.class` files - https://github.com/sbt/zinc/blob/develop/internal/compiler-bridge/src/main/scala/xsbt/API.scala#L117, whereas in Java Analyzer phase registers local classes, which should not impact the downstream.

dwijnand commented:
I have a rebase of this, but why can't I force push it? 😕

eed3si9n commented:
@dwijnand I just checked Allow edits by maintainers. Let me know if that doesn't work.

eed3si9n commented:
Actually I'm attempting to check it, but it seems like the checkbox is not sticking. I'll just give you push access to my zinc?

dwijnand committed:
Only isPickleJava when Mixed

dwijnand committed:
Mark subproject-java PENDING

With only Scala pipelining implemented, this Java-only pipelining test
is pending.

The issue lies in the fact that while scalac can consume its own .sig
early output artifacts, javac can't.

dwijnand committed:
Fix how IncrementalCallback completes cycle

mergeAndInvalidate will merge the partialAnalysis with the pruned
Analysis, so merged it twice will break it (duplicate Compilations).

Only record the generated products in the partialAnalysis - fails the
restore-classes scripted test if not.

dwijnand committed:
Buffer scripted again

dwijnand committed:
Add profiler to ExternalHooks, as a Java interface

The entry point API of Zinc, xsbti.compile.IncrementalCompiler, is
defined in compiler-interface and is, therefore, entirely in Java, only.

Therefore, in order to allow a user (e.g. an external build tool) to
define their own profiler using this interface I extracted a Java-only
interface for InvalidationProfiler and its components.

Having done that I could add it to compiler-interface's
xsbti.compile.ExternalHooks.

I've kept the original, Scala variant, interface methods (e.g. using
Scala's Iterable instead of Array or java.util.Set) for backwards
compatibility and I've created an AdaptedRunProfiler that knows how to
adapt the calls to the underlying XRunProfiler it contains.

dwijnand committed:
Close jars if the cp contains any output jars

This is to avoid the "The process cannot access the file because it is
being used by another process" issue, on Windows.

dwijnand commented:
@eed3si9n could you review this, please?  For me it's good enough to merge.  I need to look at a detail about cancelation still, but I'd prefer landing this safely first and PRing that part after.

I might also look at that `-Ypickle-write` Jason mentioned in another follow-up PR, as well as `-Ypickle-write-api-only`, but they sound like nice-to-have changes.

eed3si9n commented:
@dwijnand See my comment on https://github.com/sbt/zinc/pull/744#issuecomment-623628523. If the strategy for mixed Scala/Java subproject is to compile all the Java stuff (which is fine for now), then we need to:
1. skip Java compilation in `MixedAnalyzingCompiler#compile` and 
2. provide `compileAllJava` method to IncrementalCompilerImpl for convenience

Without it, I don't think this feature can be used by any build tool.

eed3si9n commented:
subproject-java scripted test demonstrates the need to coordinate Java compilation.
I'm fine with merging this as-is, and following up with that in a later PR.
